### PR TITLE
Returns false in case of an invalid datetime value in the datetime filter

### PR DESCRIPTION
**Pull request by DinStamou**
##### SUMMARY
This commit adds a value check in the datetime filter.
In case of an invalid datetime value as an input it will return False to the user.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
- Filter Plugin

### CODE REVIEW

File: lib/ansible/plugins/filter/core.py [hunk 1]
```diff
@@ -100,7 +100,10 @@ def to_bool(a):
 
 
 def to_datetime(string, format="%Y-%m-%d %H:%M:%S"):
-    return datetime.datetime.strptime(string, format)
+    try:
+        return datetime.datetime.strptime(string, format)
+    except ValueError:
+        return False
```
Comment (bcoca) [inline lib/ansible/plugins/filter/core.py]: this changes current default behaviour, I would reject as is, but allow a toggle to change the behaviour explicitly:

```
to_datetime(string, format, on_error=fail|ignore|etc)
```

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/filter/core.py (hunk 1 above)
Comment (DinStamou) [reply lib/ansible/plugins/filter/core.py]: I will add it like this

**Review by DinStamou — COMMENTED** [review]
@bcoca can you check this change?

**Comment by DinStamou** [comment]
@bcoca Any news on that, I think it is quite as requested as it is?

**Comment by DinStamou** [comment]
Any news on that?
@bcoca

File: lib/ansible/plugins/filter/core.py [hunk 2]
```diff
@@ -99,8 +99,18 @@ def to_bool(a):
...
+    Normal behaviour is to return datetime.
+    If toggle(fail_on_error) is True, in any value error it will fail.
+    '''
+    if fail_on_error:
+        try:
+            return datetime.datetime.strptime(string, format)
+        except ValueError:
+            return False
+    else:
+        return datetime.datetime.strptime(string, format)
```
Comment (Akasurde) [inline lib/ansible/plugins/filter/core.py:92]:
```suggestion
    try:
        return datetime.datetime.strptime(string, format)
    except ValueError:
        if fail_on_error:
            return AnsibleFilterError('Invalid value for conversion (%s)' % string)
```

File: lib/ansible/plugins/filter/core.py [hunk 3]
```diff
@@ -99,8 +99,18 @@ def to_bool(a):
...
-def to_datetime(string, format="%Y-%m-%d %H:%M:%S"):
-    return datetime.datetime.strptime(string, format)
+def to_datetime(string, format="%Y-%m-%d %H:%M:%S", fail_on_error=False):
+    '''
+    Normal behaviour is to return datetime.
+    If toggle(fail_on_error) is True, in any value error it will fail.
+    '''
+    if fail_on_error:
+        try:
+            return datetime.datetime.strptime(string, format)
```
Comment (Akasurde) [inline lib/ansible/plugins/filter/core.py:88]: Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example. Also, add testcase for this change.

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/filter/core.py (hunk 3 above)
Comment (DinStamou) [reply lib/ansible/plugins/filter/core.py:88]: Thanks a lot for the review, I added your suggestion and the changelog fragment . What kind of test is required in this case, any specific tests you could suggest me to check?

**Comment by sivel** [comment]
We've discussed this again and have decided against accepting this feature.  We do not believe that, even behind a toggle, that this behavior aligns with the goals of this plugin.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.libera.chat
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Pull request closed without merging**